### PR TITLE
doc: add happy path for PISPLinkingModel sequence diagram

### PR DESCRIPTION
**Pull request by kleyow**
Still WIP but want to get this up for feedback before I sign off for the night.

### CODE REVIEW

File: docs/sequence/PISPLinkingApi.puml [hunk 1]
```diff
@@ -0,0 +1,577 @@
...
+deactivate DFSP_TP_IN
+DFSP_LM -> DFSP_LM: DFSPBackendRequests.validateConsentRequest()
+activate DFSPAuthorizeSimulator
+DFSP_LM -> DFSPAuthorizeSimulator: ""POST /store/consentRequests/6789""
+DFSPAuthorizeSimulator -> DFSPAuthorizeSimulator: store consentRequest details
+DFSPAuthorizeSimulator -> DFSP_LM: 201 Created
+rnote right of DFSP_LM: state: consentRequestValidatedAndStored
+deactivate DFSPAuthorizeSimulator
+
+activate Switch
```
Comment (eoln) [inline docs/sequence/PISPLinkingApi.puml:92]: move to L94 and activate switch after DFSP_LM -> Switch: PUT /consentRequests/6789

File: docs/sequence/PISPLinkingApi.puml [hunk 2]
```diff
@@ -0,0 +1,577 @@
...
+    accountId: 'dfspa.username.5678',
+    actions: ['accounts.getBalance', 'accounts.transfer'],
+  }],
+  authChannels: ["WEB"],
+  callbackURI: 'pisp-app://callback...',
+  authURI: 'dfspa.com/authorize?consentRequestId=6789' // this is new
+}
+end note
+Switch --> DFSP_LM: 202 ACCEPTED
+deactivate DFSP_LM
```
Comment (eoln) [inline docs/sequence/PISPLinkingApi.puml:113]: call saveToKVS() before `deactivate DFSP_LM`

File: docs/sequence/PISPLinkingApi.puml [hunk 3]
```diff
@@ -0,0 +1,577 @@
...
+activate PISP_TP_IN
```
Comment (eoln) [inline docs/sequence/PISPLinkingApi.puml:114]: move activate after Switch -> PISP_TP_IN: PUT /consentRequests/6789

File: docs/sequence/PISPLinkingApi.puml [hunk 4]
```diff
@@ -0,0 +1,577 @@
...
+Switch ->  PISP_TP_IN: PUT /consentRequests/6789
+PISP_TP_IN --> Switch: 200 OK
+deactivate Switch
+activate PISP_LM
+PISP_TP_IN --> PISP_LM: Consent Request response recieved
+rnote right of PISP_LM: state: webAuthenticationChannelResponseRecieved
+PISP_LM --> PISP_LM: this.saveToKVS()
+deactivate PISP_TP_IN
```
Comment (eoln) [inline docs/sequence/PISPLinkingApi.puml:122]: move to L120, just before `rnote right of PISP_LM: state: webAuthenticationChannelResponseRecieved`

File: docs/sequence/PISPLinkingApi.puml [hunk 5]
```diff
@@ -0,0 +1,577 @@
...
+  }],
+  // model will add `authChannels`
+  authChannels: ["WEB", "OTP"],
+  callbackURI: 'pisp-app://callback...'
+}
+end note
+
+activate Switch
+Switch --> PISP_LM: 202 Accepted
+deactivate PISP_LM
```
Comment (eoln) [inline docs/sequence/PISPLinkingApi.puml:73]: drop `deactivate PISP_LM` to express that model is listening on PUT `/consentReqeusts/8789`

File: docs/sequence/PISPLinkingApi.puml [hunk 6]
```diff
@@ -0,0 +1,577 @@
...
+}
+end note
+deactivate PISP_TP_OUT
+deactivate PISP
+
+note over PISP, DFSPAuthorizeSimulator
+  The PISP is expected to redirect to the DFSP's `authUri` in their application to let the end user complete the web login flow.
+  Logging in successfully should provide the PISP with a secret to use in the next linking step.
+end note
+
```
Comment (eoln) [inline docs/sequence/PISPLinkingApi.puml:150]: let split the diagram here (on cases) so we can inject some text in documentation between phases

**Review by eoln — CHANGES_REQUESTED** [review]
- let use `activate` just after receiving a call, not before
- let split this big happy flow into cases (WEB/OTP) and phases into separated files, so we will be able to inject some documentation text in between

File: docs/sequence/PISPLinkingApi.puml [hunk 7]
```diff
@@ -0,0 +1,577 @@
...
+}
+end note
+
+deactivate PISP_TP_OUT
+deactivate PISP
+
+note over PISP, DFSPAuthorizeSimulator
+  The DFSP is expected to send an OTP to the end user which will be used in the next linking step.
+end note
+
```
Comment (eoln) [inline docs/sequence/PISPLinkingApi.puml:280]: let split here

File: docs/sequence/PISPLinkingApi.puml [hunk 8]
```diff
@@ -0,0 +1,577 @@
...
+== Authentication+Grant Consent+Register Credential ==
+autonumber 1 "<b>AUTHENTICATION-#</b>"
+rnote right of PISP
+PISP has obtained authToken from end-user(OTP) or through a callback(Web).
+end note
+PISP -> PISP_TP_OUT: POST /linking/authenticate
+rnote right of PISP #LightBlue
+PATCH /linking/authenticate/6789
```
Comment (eoln) [inline docs/sequence/PISPLinkingApi.puml:288]: POST or PATCH?

maybe let use POST `/linking/request-consent/{ConsentRequestId}/authenticate`

File: docs/sequence/PISPLinkingApi.puml [hunk 9]
```diff
@@ -0,0 +1,577 @@
...
+DFSP_TP_IN -> DFSP_LM: model.validateAuthToken()
+DFSP_TP_IN --> Switch: 202 Accepted
+deactivate Switch
+deactivate DFSP_TP_IN
+
+DFSP_LM -> DFSP_LM: DFSPBackendRequests.validateAuthToken()
+rnote right of DFSP_LM #LightBlue
+Do we need two backend endpoints for validating
+web authTokens and OTP authTokens? Or is a DFSP expected to
+validate both cases with one endpoint?
```
Comment (eoln) [inline docs/sequence/PISPLinkingApi.puml:332]: let propagate to the call payload the `authChannel` value which should be stored in the previous step when callback `PUT /consentReqeusts/{consentRequestId}` was called.

File: docs/sequence/PISPLinkingApi.puml [hunk 10]
```diff
@@ -0,0 +1,577 @@
...
+
+POST /validateAuthToken
+{
+  consentRequestId: '6789'
+  authToken: '124356'
```
Comment (eoln) [inline docs/sequence/PISPLinkingApi.puml:337]:
```suggestion
  authChannel: model.data.authChannel,
  authToken: '124356'
```

File: docs/sequence/PISPLinkingApi.puml (hunk 9 above)
Comment (eoln) [reply docs/sequence/PISPLinkingApi.puml:332]: when we send `authChannel` then we can easily mock simulator for proper behaviours and testing various scenarios

File: docs/sequence/PISPLinkingApi.puml [hunk 11]
```diff
@@ -0,0 +1,577 @@
...
+
+DFSP_LM -> DFSP_LM: model.getScopes()
+DFSP_LM -> DFSP_LM: Get scopes.
+rnote right of DFSP_LM #LightBlue
+Do we need to get scopes from
+- DFSPAuthorizeSimulator?
+- Using DFSPBackendRequests?
+- Can't we store the consentRequest in the Consent Request stage
+  in the model data and pull it from there instead of having to
+  make a request?
```
Comment (eoln) [inline docs/sequence/PISPLinkingApi.puml:350]: we should store results from the previous call to getScopes this data in the model state so we don't have to repeat it here.

File: docs/sequence/PISPLinkingApi.puml [hunk 12]
```diff
@@ -0,0 +1,577 @@
...
+  }],
+  credential: {
+    credentialType: "FIDO",
+    status: "PENDING",
+    payload: PublicKeyCredential
+  }
+}
+end note
+Switch --> DFSP_LM: 202 Accepted
+deactivate DFSP_LM
```
Comment (eoln) [inline docs/sequence/PISPLinkingApi.puml:469]: drop `deactivate DFSP_LM` here to express the model is waiting on `PUT /consents/1a2b3c4d` from auth service OR use `saveToKVS()`to express preserving the state

File: docs/sequence/PISPLinkingApi.puml [hunk 13]
```diff
@@ -0,0 +1,577 @@
...
+  credential: {
+    credentialType: "FIDO",
+    status: "VERIFIED",
+    payload: PublicKeyCredential
+  }
+}
+end note
+
+Switch --> AUTH: 200 OK
+deactivate AUTH
```
Comment (eoln) [inline docs/sequence/PISPLinkingApi.puml:497]: let drop `deactivate AUTH` here and `activate AUTH` in L509 to express AUTH is sill processing

**Review by eoln — CHANGES_REQUESTED** [review]
more comments

**Comment by kleyow** [comment]
> * let use `activate` just after receiving a call, not before
> * let split this big happy flow into cases (WEB/OTP) and phases into separated files, so we will be able to inject some documentation text in between

I was actually considering using `autoactivate` or `++` or `--` notation and wanted to know other peoples thoughts are on that.

**Comment by eoln** [comment]
let use autoactivate if it fits you, I wouldn't use `++` or `--` because they are too cryptic

**Comment by kleyow** [comment]
autoactivate turned out to not be very intuiative so didn't bother to use it. addressed comments.

File: docs/sequence/PISPLinkingOTPApi.puml [hunk 14]
```diff
@@ -0,0 +1,431 @@
...
+Switch --> PISP_LM: 202 Accepted
+deactivate PISP_LM
+Switch -> DFSP_TP_IN: POST /consentRequests
+activate DFSP_TP_IN
+
+DFSP_TP_IN -> DFSP_TP_IN: const model = await create()
+activate DFSP_LM
+rnote right of DFSP_TP_IN: state: start
+DFSP_TP_IN -> DFSP_LM: model.requestConsent()
+DFSP_TP_IN --> Switch: 202 Accepted
```
Comment (eoln) [inline docs/sequence/PISPLinkingOTPApi.puml:80]: move to L75 -> there is a need to send 202 back to the switch asap

File: docs/sequence/PISPLinkingOTPApi.puml [hunk 15]
```diff
@@ -0,0 +1,431 @@
...
+end note
+
+activate Switch
+Switch --> PISP_LM: 202 Accepted
+deactivate PISP_LM
+Switch -> DFSP_TP_IN: POST /consentRequests
+activate DFSP_TP_IN
+
+DFSP_TP_IN -> DFSP_TP_IN: const model = await create()
+activate DFSP_LM
```
Comment (eoln) [inline docs/sequence/PISPLinkingOTPApi.puml:77]: too early activation of DFSP_LM - move to L80

File: docs/sequence/PISPLinkingOTPApi.puml [hunk 16]
```diff
@@ -0,0 +1,431 @@
...
+Switch --> PISP_LM: 202 Accepted
+deactivate PISP_LM
+Switch -> DFSP_TP_IN: PATCH /consentRequests/6789
+activate DFSP_TP_IN
+
+DFSP_TP_IN -> DFSP_TP_IN: const model = await loadFromKVS({key: 6789})
+activate DFSP_LM
+rnote right of DFSP_TP_IN: state: consentRequestValidatedAndStored
+DFSP_TP_IN -> DFSP_LM: model.validateAuthToken()
+DFSP_TP_IN --> Switch: 202 Accepted
```
Comment (eoln) [inline docs/sequence/PISPLinkingOTPApi.puml:195]: 202 asap to switch -> move to L190

File: docs/sequence/PISPLinkingOTPApi.puml [hunk 17]
```diff
@@ -0,0 +1,431 @@
...
+end note
+
+activate Switch
+Switch --> PISP_LM: 202 Accepted
+deactivate PISP_LM
+Switch -> DFSP_TP_IN: PATCH /consentRequests/6789
+activate DFSP_TP_IN
+
+DFSP_TP_IN -> DFSP_TP_IN: const model = await loadFromKVS({key: 6789})
+activate DFSP_LM
```
Comment (eoln) [inline docs/sequence/PISPLinkingOTPApi.puml:192]: to early activation, move to L195, after `model.validateAuthToken()`

File: docs/sequence/PISPLinkingOTPApi.puml [hunk 18]
```diff
@@ -0,0 +1,431 @@
...
+PISP_TP_IN -> PISP_LM: model.registerCredential()
+PISP_TP_IN --> Switch: 202 Accepted
+deactivate PISP_TP_IN
+deactivate Switch
+rnote right of PISP_LM: state: consentRecieved
+rnote right of PISP_LM #LightBlue
+It's important to save the model with the consentId from this point onwards!
+Requests will not contain a consentRequestId in the upcoming requests.
+This may require changing `PersistentModel` to have a secondary key it can
+store the model with.
```
Comment (eoln) [inline docs/sequence/PISPLinkingOTPApi.puml:254]: I see two possible solutions:
- we can create the second instance of the model using existing state data but with a new key and call saveToKVS
- or we can refactor save/load methods and add an optional parameters to specify special/new key.

File: docs/sequence/PISPLinkingOTPApi.puml [hunk 19]
```diff
@@ -0,0 +1,431 @@
...
+DFSP_TP_IN -> DFSP_TP_IN: participantResponseRecieved()
+rnote right of DFSP_TP_IN: state: participantsResponseRecieved
+DFSP_TP_IN -> DFSP_LM: Participant response recieved
+DFSP_TP_IN --> ALS: 200 Accepted
+deactivate ALS
+deactivate DFSP_TP_IN
+rnote right of DFSP_LM: state: consentRegisteredAndValidated
+DFSP_LM -> DFSP_LM: model.finalizeConsent()
+
+loop for each scope in ""Consents.scopes""
```
Comment (eoln) [inline docs/sequence/PISPLinkingOTPApi.puml:381]: maybe it would be better to consider bulk update instead of multiple calls - @lewisdaly ???

**Review by eoln — CHANGES_REQUESTED** [review]
- inbound server should send 202 back to switch ASAP for both diagrams

**Review by eoln — APPROVED** [review]

**Pull request merged**